src/ephy-sync-crypto.c

```diff
@@ -0,0 +1,952 @@
+  digest = ephy_sync_crypto_decode_hex (digest_hex);
+  mac = g_base64_encode (digest, g_checksum_type_get_length (G_CHECKSUM_SHA256));
+
+  g_free (normalized);
+  g_free (digest_hex);
+  g_free (digest);
+
+  return mac;
+}
+
+static char *
+ephy_sync_crypto_append_to_header (char       *header,
+                                   const char *name,
+                                   char       *value)
+{
+  char *new_header;
+  char *tmp;
+
+  g_return_val_if_fail (header != NULL, NULL);
+  g_return_val_if_fail (name != NULL, NULL);
+  g_return_val_if_fail (value != NULL, NULL);
+
+  tmp = header;
+  new_header = g_strconcat (header, ", ", name, "=\"", value, "\"", NULL);
+  g_free (tmp);
+
+  return new_header;
+}
+
+static void
+ephy_sync_crypto_hkdf (guint8 *in,
+                       gsize   in_len,
+                       guint8 *salt,
+                       gsize   salt_len,
+                       guint8 *info,
+                       gsize   info_len,
+                       guint8 *out,
+                       gsize   out_len)
+{
+  char *prk_hex;
+  char *tmp_hex;
+  guint8 *tmp;
+  guint8 *prk;
+  guint8 *out_full;
+  guint8 *data;
+  guint8 counter;
+  gsize hash_len;
+  gsize data_len;
+  gsize n;
+
+  g_return_if_fail (in != NULL);
+  g_return_if_fail (info != NULL);
+  g_return_if_fail (out != NULL);
+
+  hash_len = g_checksum_type_get_length (G_CHECKSUM_SHA256);
+  g_assert (out_len <= hash_len * 255);
+
+  /* If salt value was not provided, use an array of hash_len zeros */
+  if (salt == NULL) {
+    salt = g_malloc0 (hash_len);
+    salt_len = hash_len;
+  }
+
+  /* Step 1: Extract */
+  prk_hex = g_compute_hmac_for_data (G_CHECKSUM_SHA256, salt, salt_len, in, in_len);
+  prk = ephy_sync_crypto_decode_hex (prk_hex);
+
+  /* Step 2: Expand */
+  counter = 1;
+  n = (out_len + hash_len - 1) / hash_len;
+  out_full = g_malloc (n * hash_len);
+
+  for (gsize i = 0; i < n; i++, counter++) {
+    if (i == 0) {
+      data = ephy_sync_utils_concatenate_bytes (info, info_len, &counter, 1, NULL);
+      data_len = info_len + 1;
+    } else {
+      data = ephy_sync_utils_concatenate_bytes (out_full + (i - 1) * hash_len, hash_len,
+                                                info, info_len, &counter, 1,
+                                                NULL);
+      data_len = hash_len + info_len + 1;
+    }
+
+    tmp_hex = g_compute_hmac_for_data (G_CHECKSUM_SHA256, prk, hash_len, data, data_len);
+    tmp = ephy_sync_crypto_decode_hex (tmp_hex);
+    memcpy (out_full + i * hash_len, tmp, hash_len);
+
+    g_free (data);
+    g_free (tmp);
+    g_free (tmp_hex);
+  }
+
+  memcpy (out, out_full, out_len);
+
+  g_free (prk_hex);
+  g_free (salt);
+  g_free (prk);
+  g_free (out_full);
+}
+
+static void
+ephy_sync_crypto_random_gen (void   *ctx,
+                             gsize   length,
+                             guint8 *dst)
+{
+  for (gsize i = 0; i < length; i++)
+    dst[i] = g_random_int ();
```

Comment:
Nonononono, this API is documented to be unsuitable for crypto. Didn't you have code for reading /dev/urandom at one point? That's what you'll want to use here.